rfcs/20190829-tfx-container-component-execution.md

```diff
@@ -0,0 +1,555 @@
+    steps in a TFX pipeline.
+*   User wants to use non-python code like R as one of the steps in a TFX
+    pipeline.
+*   User wants to have an isolated python environment for their component code.
+
+This RFC is a follow-up design for
+[Container-based Component RFC](https://github.com/tensorflow/community/pull/146).
+This design defines how to execute the container spec as part of a TFX pipeline.
+The execution may occurs in local docker or in a remote kubernetes cluster.
+
+### Existing solutions
+
+#### Kubeflow Pipeline (KFP) ContainerOp
+
+Today, KFP’s ContainerOp leverages
+[Argo container template API](https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/workflow_types.go)
+to launch user’s container in a k8s pod. Argo as the orchestrator controls when
+to launch the POD and it uses a sidecar container to report output files back
+and wait for user’s container to complete. We are not proposing to use Argo API
+because of the following reasons:
+
+*   Argo’s API is orchestrator-specific and cannot be ported to Airflow or local
+    runners.
+*   Argo’s API doesn’t provide an extensible way to run custom code before and
+    after POD API which is critical to support metadata tracking and caching
+    features.
+*   Argo doesn’t provide an easy way to recover from user’s transient errors,
+    which is critical in production workload.
+
+#### Airflow k8s pod operator
+
+Airflow supports to launch a k8s pod by an
+[operator](https://github.com/apache/airflow/blob/master/airflow/contrib/operators/kubernetes_pod_operator.py).
+This approach is closer to what we are proposing in the document. However, we
+cannot directly use the operator because:
+
+*   Airflow operator requires to be run inside an Airflow pipeline which is not
+    the case for local and KF runners.
+*   Airflow operator exposes a subset of POD’s API, where we want to expose the
+    full pod spec to the user.
+*   Airflow operator doesn’t provide a reliable way to retry user’s container
+    and recover from transient errors.
+*   Airflow does not support initializing operator inside another operator.
+    Going back to use multiple Airflow operators for a component is a regression
+    now that we have `BaseComponentLauncher` ready.
+
+## Proposed Design
+
+### TLDR
+
+We propose to solve the above problems by the following design.
+
+*   Define container as an executor spec.
+*   Launch container by component launcher in either local docker or k8s pod.
+*   Use platform config to specify platform specific settings like k8s pod
+    config.
+
+The proposed solution has following parts:
+
+*   Extensible `ExecutorSpec` concept which can support container as an
+    executor.
+*   Extensible `BaseComponentLauncher` concept to support pluggable component
+    launchers in tfx runner.
+    *   `DockerComponentLauncher` which launches `ExecutorContainerSpec` in
+        docker environment.
```

Comment:
Although initially targeted for "Docker", I would suggest changing this to something like "Any OCI container system, which includes Docker".  In other words, have a goal to work with any OCI container technology.

Reply:
sg. The spec is not specific to docker. We just provide an implementation for docker.